include/boost/url/impl/url.ipp

```diff
@@ -2046,22 +2046,92 @@ url&
 url::
 normalize()
 {
-#if 0
-    // scheme
-    {
-        char* it = s_
-            + offset(id_scheme);
-        auto last = it + offset(id_user);
-        if(it < last)
+    char* dest = s_;
+    char buf{'\0'};
+    int all_ids[] = {id_scheme, id_user, id_pass,
+                      id_host,  id_port, id_path,
+                      id_query, id_frag};
+    int next_id_offset = offset(id_scheme);
+    for (auto id: all_ids)
+    {
+        char* it = s_ + next_id_offset;
+        char* end = s_ + offset(id + 1);
+        while (it < end)
+        {
+            if (*it != '%')
+            {
+                *dest = *it;
+                ++it;
+                ++dest;
+                continue;
+            }
+            if (end - it < 3)
+                break;
+
+            // decode unreserved octets
+            pct_decode_unchecked(
+                &buf,
+                &buf + 1,
+                string_view(it, 3));
+            if (unreserved_chars(buf))
+            {
+                *dest = buf;
+                it += 3;
+                ++dest;
+                continue;
+            }
+
+            // uppercase percent-encoding triplets
+            ++it;
+            *it = grammar::ascii_toupper(*it);
+            ++it;
+            *it = grammar::ascii_toupper(*it);
+            ++it;
+            dest += 3;
+        }
+        // update offset of next component
+        next_id_offset = offset_[id + 1];
+        std::size_t new_offset = dest - s_;
+        std::size_t diff =
+            offset_[id + 1] - new_offset;
+        offset_[id + 1] -= diff;
+    }
+    *dest = '\0';
+    offset_[id_end] = dest - s_;
+
+    // scheme and host to lowercase
+    int to_lower_ids[] = {id_scheme, id_host};
+    for (auto id: to_lower_ids)
+    {
+        char* it = s_ + offset(id);
+        char* end = s_ + offset(id + 1);
+        if(it < end)
         {
-            --last;
-            while(it != last)
+            while(it < end)
             {
-                // to upper
+                if (*it != '%')
+                {
+                    *it = grammar::ascii_tolower(
+                        *it);
+                    ++it;
+                    continue;
+                }
+                it += 3;
             }
         }
     }
-#endif
+
+    // Path-segments are normalized with remove_dot_segments
+    string_view p = encoded_path();
+    char* p_dest = s_ + offset(id_path);
+    char* p_end = s_ + offset(id_path + 1);
+    std::size_t pn = p.size();
+    std::size_t n = detail::remove_dot_segments(
+        p_dest, p_end, p, is_path_absolute());
+    if (n != pn) {
+        resize_impl(id_path, n);
```

Comment:
I think we still need to update some url_view pre-processed data here.

Reply:
I changed it, but just to remove some redundant conditionals. It's the same logic. 

About the pre-processed data, I'm still working on it.